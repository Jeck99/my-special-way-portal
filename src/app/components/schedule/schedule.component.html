<table class="mat-table schedule-table" role="grid">
  <thead>
    <tr class="mat-header-row" role="row">
      <th class="mat-header-cell cdk-column-position mat-column-position" role="columnheader" colspan="2">שעה</th>
      <th class="mat-header-cell cdk-column-position mat-column-position" role="columnheader" *ngFor="let day of daysLabels">{{day}}</th>
    </tr>
  </thead>
  <tbody>
    <tr class="mat-row" role="row" *ngFor="let row of schedule; index as hourIndex">
      <td class="mat-cell cdk-column-position mat-column-position" role="gridcell">{{hourIndex + 1}}</td>
      <td class="mat-cell cdk-column-position mat-column-position" role="gridcell">{{hoursLabels[hourIndex]}}</td>
      <td 
        class="mat-cell cdk-column-position mat-column-position"
        *ngFor="let cell of row; index as dayIndex" 
        role="gridcell" 
        (click)="timeSlotClicked.emit({hourIndex: hourIndex, dayIndex: dayIndex})"
      >
        <div *ngIf="cell" class="timeslot-full">
          {{cell.title}}
        </div>
        <div *ngIf="!cell" class="empty-timeslot">
          <i class="material-icons" >
            add
          </i>
        </div>
      </td>
    </tr>
  </tbody>
</table>
