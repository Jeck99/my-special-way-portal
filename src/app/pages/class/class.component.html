<div class="main-content">
  <div class="container-fluid">
    <div class="header-menu-container msw-table-tools-header">
      <div class="header-table-count">
        <div>
          <span class="header-table-count-label">כיתות</span>
          <span class="header-table-count-number">({{dataSource.data.length}})</span>
        </div>
      </div>
      <div class="example-header msw-table-tools-header">
        <button mat-raised-button color="primary" [routerLink]="['/class', '_new_']">
          <i class="material-icons">
            add
          </i>
          הוספת כיתה חדשה
        </button>
      </div>
    </div>
    <div class="row">

      <div class="col-lg-12 col-md-12 col-sm-12">
        <mat-table #table [dataSource]="dataSource" class="msw-table"
                   matSort matSortActive="name" matSortDisableClear matSortDirection="asc">

          <!-- Number Column -->
          <ng-container matColumnDef="classname">
            <mat-header-cell *matHeaderCellDef>שם כיתה</mat-header-cell>
            <mat-cell *matCellDef="let row" (click)="navigate(row._id)">{{ row.name }}</mat-cell>
          </ng-container>

          <!-- Title Column -->
          <ng-container matColumnDef="level">
            <mat-header-cell *matHeaderCellDef>כיתה</mat-header-cell>
            <mat-cell *matCellDef="let row" (click)="navigate(row._id)">{{ scheduleService.grades[row.grade] }}</mat-cell>
          </ng-container>

          <!--&lt;!&ndash; State Column &ndash;&gt;-->
          <ng-container matColumnDef="deleteClass">-->
            <mat-header-cell *matHeaderCellDef></mat-header-cell>
            <mat-cell *matCellDef="let row">
              <button mat-icon-button color="primary" (click)="deleteClass(row._id, row.name, row.level, row.students.length)"><i class="material-icons"> delete </i></button>
            </mat-cell>
          </ng-container>

          <mat-header-row *matHeaderRowDef="displayedColumns"></mat-header-row>
          <mat-row *matRowDef="let row; columns: displayedColumns"></mat-row>
        </mat-table>
      </div>
    </div>
  </div>
</div>


<!-- Copyright 2018 Google Inc. All Rights Reserved.
    Use of this source code is governed by an MIT-style license that
    can be found in the LICENSE file at http://angular.io/license -->
