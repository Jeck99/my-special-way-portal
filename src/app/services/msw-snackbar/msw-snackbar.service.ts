import { Injectable } from '@angular/core';
import { MatSnackBar } from '@angular/material';

@Injectable()
export class MSWSnackbar {
  constructor(private snackbar: MatSnackBar) {}

  public displayTimedMessage(message: string, duration: number = 3000) {
    this.snackbar.open(message, undefined, { duration });
  }
}
