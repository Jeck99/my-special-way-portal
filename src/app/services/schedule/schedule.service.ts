import { Injectable } from '@angular/core';
import { TimeSlot } from '../../models/timeslot.model';

@Injectable()
export class ScheduleService {
  levels = ['א', 'ב', 'ג', 'ד', 'ה', 'ו'];
  daysLabels = ['ראשון', 'שני', 'שלישי', 'רביעי', 'חמישי', 'שישי'];
  hoursLabels = [
    '07:30 - 08:00',
    '08:00 - 08:50',
    '08:50 - 09:35',
    '09:35 - 10:00',
    '10:00 - 10:30',
    '10:30 - 11:15',
    '11:15 - 12:00',
    '12:00 - 12:45',
    '12:45 - 13:15',
    '13:15 - 13:45',
    '13:45 - 14:30',
    '14:30 - 15:15',
    '15:15 - 16:00',
    '16:00 - 16:30',
    '16:30 - 16:45',
  ];

  buildScheduleFromTimeslots(
    hoursCount: number,
    daysCount: number,
    timeslots: TimeSlot[],
  ): TimeSlot[][] {
    const schedule: TimeSlot[][] = [];

    for (let hourIndex = 0; hourIndex < hoursCount; hourIndex++) {
      schedule[hourIndex] = new Array(daysCount);

      for (let dayIndex = 0; dayIndex < daysCount; dayIndex++) {
        const timeslot = timeslots.find((t) => t.index === `${hourIndex}${dayIndex}`);
        const newTimeSlot: TimeSlot = {
          index: `${hourIndex}${dayIndex}`,
        };
        if (timeslot) {
          if (timeslot.location) {
            newTimeSlot.location = timeslot.location;
          }
          if (timeslot.lesson) {
            newTimeSlot.lesson = timeslot.lesson;
          }
        }
        schedule[hourIndex][dayIndex] = newTimeSlot;
      }
    }

    return schedule;
  }
}
