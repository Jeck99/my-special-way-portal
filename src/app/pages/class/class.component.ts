import {Component, OnInit, ElementRef, ViewChild} from '@angular/core';
import 'rxjs/add/observable/merge';
import 'rxjs/add/observable/fromEvent';
import 'rxjs/add/operator/map';
import 'rxjs/add/operator/debounceTime';
import 'rxjs/add/operator/distinctUntilChanged';
import { MatTableDataSource, MatSort, MatDialog } from '@angular/material';
import { ClassService } from './services/class.graphql.service';
import { AddClassDialogComponent } from './dialogs/add/add-class.dialog';
import { Class } from '../../models/class.model';
import { DeleteClassDialogComponent } from './dialogs/delete/delete-class.dialog';
import * as _ from 'lodash';
import { UpdateClassDialogComponent } from './dialogs/update/update-class.dialog';

@Component({
  selector: 'app-grade',
  templateUrl: './class.component.html',
  styleUrls: ['./class.component.scss'],
})
export class ClassComponent implements OnInit {

  displayedColumns = ['classname', 'level', 'editDetails', 'deleteClass'];
  dataSource = new MatTableDataSource<Class>();
  resultsLength = 0;

  @ViewChild(MatSort) sort: MatSort;
  @ViewChild('table') table: ElementRef;

  constructor(
    private classService: ClassService,
    public dialog: MatDialog,
  ) { }

  async ngOnInit() {
    await this.populateDatasource();
    this.dataSource.sort = this.sort;
  }

  private async populateDatasource() {
    try {
      this.dataSource.data = await this.classService.getAllClasses();
    } catch (error) {
      // TODO: implement error handling on UI
      console.error('Error handling not implemented');
      throw error;
    }
  }

  applyFilter(filterValue: string) {
    filterValue = filterValue.trim(); // Remove whitespace
    filterValue = filterValue.toLowerCase(); // Datasource defaults to lowercase matches
    this.dataSource.filter = filterValue;
  }

  addNewClass() {
    const dialogRef = this.dialog.open(AddClassDialogComponent, {
      data: { class: Class },
      height: '310px',
      width: '320px',
    });
    dialogRef.afterClosed().subscribe((data) => {
      if (data) {
        const newClass: Class = this._createNewClass(data);
        this.classService.create(newClass)
          .then(() => {
            this.dataSource.data.push(newClass);
          });
      }
    });
  }

  deleteClass(_id: string, name: string, level: string) {
    const dialogRef = this.dialog.open(DeleteClassDialogComponent, {
      data: {_id, name, level},
    });

    dialogRef.afterClosed().subscribe((result) => {
      if (result === true) {
        this.classService.delete(_id)
          .then(() => {
            const index = _.findIndex(this.dataSource.data, (user) =>  user._id === _id);
            this.dataSource.data.splice(index, 1);
          });
      }
    });
  }

  editClass(_id: string, name: string, level: string) {
    const dialogRef = this.dialog.open(UpdateClassDialogComponent, {
      data: {_id, name, level},
      height: '310px',
      width: '320px',
    });

    dialogRef.afterClosed().subscribe((result) => {
      if (result) {
        const relevantClass = _.find(this.dataSource.data, {_id});
        const tempClass = _.assign({}, relevantClass, result);

        this.classService.update(tempClass)
          .then((data) => {
            const index = _.findIndex(this.dataSource.data, (user) => user._id === _id);
            this.dataSource.data[index] = _.assign({}, relevantClass, result);
          });
      }
    });
  }

  _createNewClass(classData: any): Class {
    const _class: Class = new Class();
    _class.name = classData.name;
    _class.level = classData.level;
    _class.number = 1;
    return _class;
  }
}
