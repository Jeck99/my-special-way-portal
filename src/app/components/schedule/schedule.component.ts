import { Component, Input, Output, EventEmitter, ChangeDetectionStrategy, OnInit } from '@angular/core';
import { TimeSlot } from '../../models/timeslot.model';
import { MSWSnackbar } from '../../services/msw-snackbar/msw-snackbar.service';
import { FormClassData } from '../../models/FormClassData.model';

export interface TimeSlotIndexes {
  hourIndex: number;
  dayIndex: number;
}

@Component({
  selector: 'app-schedule',
  templateUrl: './schedule.component.html',
  styleUrls: ['./schedule.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class ScheduleComponent implements OnInit {
  @Input()
  schedule: TimeSlot[][];
  @Input()
  daysLabels: string[];
  @Input()
  hoursLabels: string[];
  @Input()
  selectedClassData: FormClassData;
  @Output()
  timeSlotClicked: EventEmitter<TimeSlotIndexes> = new EventEmitter();

  constructor(private mswSnackbar: MSWSnackbar) {}

  ngOnInit() {
    this.timeSlotClicked.subscribe((val) => {
      if (!val.selectedClassData.class) {
        this.mswSnackbar.displayTimedMessage('כיתה הינו שדה חובה');
        return;
      }
      if (!val.selectedClassData.grade) {
        this.mswSnackbar.displayTimedMessage('שכבה הינו שדה חובה');
        return;
      }
    });
  }
}
