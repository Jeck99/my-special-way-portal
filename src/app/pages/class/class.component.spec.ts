import { ClassComponent } from './class.component';
import { TestBed } from '@angular/core/testing';
import { NO_ERRORS_SCHEMA } from '@angular/core';
import { MatHeaderRow, MatRowDef, MatHeaderRowDef, MatSort, MatDialog } from '@angular/material';
import { ClassService } from './services/class.graphql.service';
import { classTestData } from '../../../mocks/assets/classes.mock';
import { Overlay, ScrollStrategyOptions, ScrollDispatcher, OverlayKeyboardDispatcher,
  OverlayPositionBuilder, OverlayContainer, ViewportRuler } from '@angular/cdk/overlay';
import { Platform } from '@angular/cdk/platform';
import { Observable } from 'rxjs-compat';
import { ScheduleService } from '../../services/schedule/schedule.service';
import {MSWSnackbar} from '../../services/msw-snackbar/msw-snackbar.service';
import {Router} from '@angular/router';

describe('class component', () => {
  let classServiceMock: Partial<ClassService>;
  let classDialogMock: Partial<MatDialog>;
  let scheduleServiceMock: Partial<ScheduleService>;
  let snackbarMock: Partial<MSWSnackbar>;
  let routeMock: Partial<Router>;

  beforeEach(async () => {

    classServiceMock = {
      getAllClasses: jest.fn(),
      delete: jest.fn(),
      create: jest.fn(),
      update: jest.fn(),
    };

    classDialogMock = {
      open: jest.fn().mockReturnValue({
        afterClosed: jest.fn().mockReturnValue(Observable.of(true)),
      }),
    };

    snackbarMock = {
      displayTimedMessage: jest.fn(),
    };

    scheduleServiceMock = {
      grades: {
        a: 'א',
        b: 'ב',
        c: 'ג',
        d: 'ד',
        e: 'ה',
        f: 'ו',
      },
    };

    TestBed.configureTestingModule({
      imports: [
      ],
      declarations: [
        ClassComponent,
        MatHeaderRow,
        MatRowDef,
        MatHeaderRowDef,
        MatSort,
      ],
      providers: [
        { provide: MatDialog, useValue: classDialogMock },
        { provide: ClassService, useValue: classServiceMock },
        { provide: ScheduleService, useValue: scheduleServiceMock},
        { provide: MSWSnackbar, useValue: snackbarMock}
        { provide: Router, useValue: routeMock },
        Overlay,
        ScrollStrategyOptions,
        ScrollDispatcher,
        Platform,
        ViewportRuler,
        OverlayContainer,
        OverlayPositionBuilder,
        OverlayKeyboardDispatcher,
      ],
      schemas: [NO_ERRORS_SCHEMA],
    });

  });

  it('should render component as described in snapshot', () => {
    const fixture = TestBed.createComponent(ClassComponent);
    expect(fixture).toMatchSnapshot();
  });

  it('should open dialog when calling deleteClass function when no students assigned to the class', () => {
    (classServiceMock.delete as jest.Mock).mockImplementationOnce(
      () => {return Promise.resolve(1);
    });
    const fixture = TestBed.createComponent(ClassComponent);
    fixture.componentInstance.deleteClass('123', 'ddd', 'ddd', 0);
    const DialogMock = TestBed.get(MatDialog);
    expect(DialogMock.open).toHaveBeenCalled();
  });

  it('should not open a dialog when calling deleteClass function when students are assigned to the class', () => {
    (classServiceMock.delete as jest.Mock).mockImplementationOnce(
      () => {return Promise.resolve(0);
      });
    const fixture = TestBed.createComponent(ClassComponent);
    fixture.componentInstance.deleteClass('123', 'ddd', 'ddd', 1);
    const DialogMock = TestBed.get(MatDialog);
    expect(DialogMock.open).not.toHaveBeenCalled();
  });

  it('should open a snackbar error message when calling deleteClass function when students are assigned to the class', () => {
    (classServiceMock.delete as jest.Mock).mockImplementationOnce(
      () => {return Promise.resolve(0);
      });
    const fixture = TestBed.createComponent(ClassComponent);
    fixture.componentInstance.deleteClass('123', 'ddd', 'ddd', 1);
    const SnackbarMock = TestBed.get(MSWSnackbar);
    expect(SnackbarMock.displayTimedMessage).toHaveBeenCalled();
  });

  it('should load classes from service on page load ', () => {
    (classServiceMock.getAllClasses as jest.Mock).mockImplementationOnce(
      () => {return Promise.resolve(Promise.resolve(classTestData.classes));
    });
    const fixture = TestBed.createComponent(ClassComponent);
    fixture.detectChanges();
    const DialogMock = TestBed.get(ClassService);
    expect(DialogMock.getAllClasses).toHaveBeenCalled();
  });

  it('should load correct number of classes', async () => {
    (classServiceMock.getAllClasses as jest.Mock).mockImplementationOnce(
      () => {return Promise.resolve(classTestData.classes);
    });
    const fixture = TestBed.createComponent(ClassComponent);
    fixture.detectChanges();
    await fixture.whenRenderingDone();
    expect(fixture.componentInstance.dataSource.data.length).toEqual(23);
  });
  it('should load zero classes in case of promise reject', async () => {
    (classServiceMock.getAllClasses as jest.Mock).mockImplementationOnce(
      () => {return Promise.resolve(Promise.reject());
    });
    const fixture = TestBed.createComponent(ClassComponent);
    fixture.detectChanges();
    await fixture.whenRenderingDone();
    expect(fixture.componentInstance.dataSource.data.length).toEqual(0);
  });
});
