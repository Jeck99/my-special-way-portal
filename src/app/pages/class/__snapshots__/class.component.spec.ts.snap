// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`class component should render component as described in snapshot 1`] = `
<app-grade
  classService={[Function Object]}
  dataSource={[Function MatTableDataSource]}
  dialog={[Function Object]}
  displayedColumns={[Function Array]}
  mswSnackbar={[Function Object]}
  resultsLength="0"
  router="undefined"
  scheduleService={[Function Object]}
  sort={[Function MatSort]}
  table={[Function ElementRef]}
>
  <div
    class="main-content msw-class-container"
    data-test-id="classes-page"
  >
    <div
      class="container-fluid"
    >
      <div
        class="header-menu-container msw-table-tools-header"
      >
        <div
          class="header-table-count"
        >
          <div>
            <span
              class="header-table-count-label"
            >
              כיתות
            </span>
            <span
              class="header-table-count-number"
            >
              (
            </span>
          </div>
        </div>
        <div
          class="example-header msw-table-tools-header"
        >
          <button
            color="primary"
            mat-raised-button=""
          >
            <i
              class="material-icons"
            >
               add 
            </i>
             הוסף כיתה חדשה 
          </button>
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="col-lg-12 col-md-12 col-sm-12"
        >
          <mat-table
            class="msw-table"
            matsort=""
            matsortactive="name"
            matsortdirection="asc"
            matsortdisableclear=""
          >
            
            
            
            
            
            
            
            --&gt; 
            
            
            
            
          </mat-table>
        </div>
      </div>
    </div>
  </div>
</app-grade>
`;
